### ISSUE TITLE
Intel iwlwifi stopped working with kernel 4.13

### ISSUE THREAD

**Issue by bendlas** [NixOS/nixpkgs#29806]
## Issue description

When booting with a `4.13` kernel, my wifi card (`AC 8265, REV=0x230`) fails to get its firmware loaded:

```
Sep 26 14:26:54 lenix kernel: iwlwifi 0000:03:00.0: enabling device (0000 -> 0002)
Sep 26 14:26:54 lenix kernel: iwlwifi 0000:03:00.0: Direct firmware load for iwlwifi-8265-33.ucode failed with error -2
Sep 26 14:26:54 lenix kernel: iwlwifi 0000:03:00.0: Direct firmware load for iwlwifi-8265-32.ucode failed with error -2
Sep 26 14:26:54 lenix kernel: iwlwifi 0000:03:00.0: Direct firmware load for iwlwifi-8265-31.ucode failed with error -2
Sep 26 14:26:54 lenix kernel: iwlwifi 0000:03:00.0: Direct firmware load for iwlwifi-8265-30.ucode failed with error -2
Sep 26 14:26:54 lenix kernel: iwlwifi 0000:03:00.0: Direct firmware load for iwlwifi-8265-29.ucode failed with error -2
Sep 26 14:26:54 lenix kernel: iwlwifi 0000:03:00.0: Direct firmware load for iwlwifi-8265-28.ucode failed with error -2
Sep 26 14:26:54 lenix kernel: iwlwifi 0000:03:00.0: Direct firmware load for iwlwifi-8265-27.ucode failed with error -2
Sep 26 14:26:54 lenix kernel: iwlwifi 0000:03:00.0: Direct firmware load for iwlwifi-8265-26.ucode failed with error -2
Sep 26 14:26:54 lenix kernel: iwlwifi 0000:03:00.0: Direct firmware load for iwlwifi-8265-25.ucode failed with error -2
Sep 26 14:26:54 lenix kernel: iwlwifi 0000:03:00.0: Direct firmware load for iwlwifi-8265-24.ucode failed with error -2
Sep 26 14:26:54 lenix kernel: iwlwifi 0000:03:00.0: Direct firmware load for iwlwifi-8265-23.ucode failed with error -2
Sep 26 14:26:54 lenix kernel: iwlwifi 0000:03:00.0: Direct firmware load for iwlwifi-8265-22.ucode failed with error -2
Sep 26 14:26:54 lenix kernel: iwlwifi 0000:03:00.0: no suitable firmware found!
Sep 26 14:26:54 lenix kernel: iwlwifi 0000:03:00.0: minimum version required: iwlwifi-8265-22
Sep 26 14:26:54 lenix kernel: iwlwifi 0000:03:00.0: maximum version supported: iwlwifi-8265-33
Sep 26 14:26:54 lenix kernel: iwlwifi 0000:03:00.0: check git://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git
```
With `4.12`, the firmware load works:

```
Sep 26 14:27:42 lenix kernel: iwlwifi 0000:03:00.0: enabling device (0000 -> 0002)
Sep 26 14:27:42 lenix kernel: iwlwifi 0000:03:00.0: Direct firmware load for iwlwifi-8265-30.ucode failed with error -2
Sep 26 14:27:42 lenix kernel: iwlwifi 0000:03:00.0: Direct firmware load for iwlwifi-8265-29.ucode failed with error -2
Sep 26 14:27:42 lenix kernel: iwlwifi 0000:03:00.0: Direct firmware load for iwlwifi-8265-28.ucode failed with error -2
Sep 26 14:27:42 lenix kernel: iwlwifi 0000:03:00.0: capa flags index 3 larger than supported by driver
Sep 26 14:27:42 lenix kernel: iwlwifi 0000:03:00.0: loaded firmware version 27.541033.0 op_mode iwlmvm
```

@fpletz I appreciate your previous work on this. Do you think our build recipe (`pkgs/os-specific/linux/firmware/firmware-linux-nonfree`) could be changed to only depend on upstream sources + patches that we maintain within nixpkgs?

**Comment by fpletz** [comment]
@bendlas I totally agree! I just continued making the repository available and merge the firmware repos together like wkennington did.

Do you want to work on this? We need this for 17.09 which we will release in the next few days. :)

**Comment by bendlas** [comment]
I might give it a shot, but can you tell me a bit about it?

- Does it need to be a merged repository?
  i.e. could we also build `firmware` and `iwlwifi` separately?
- Are there any patches in there, that are from neither upstream?

also calling on @wkennington in case you're still interested in helping with this ..

**Comment by fpletz** [comment]
Should've explained what to do, sorry. :)

Only these tho repos are needed:
 * https://git.kernel.org/pub/scm/linux/kernel/git/iwlwifi/linux-firmware.git
 * https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git

These two shouldn't be needed to be built together, just linked togethe in one directory like we do with the `hardware.firmware` NixOS option. But at least for 17.09 we should install the contents of both repos with `firmwareLinuxNonfree` because that's what we've done before.

At some point we removed broken firmware but this was also removed upstream shortly after. So there should be no changes compared to the upstream repos.

**Comment by fpletz** [comment]
I don't understand why firmware loading fails for you as we have the following firmware files in `firmwareLinuxNonfree`:

```
iwlwifi-8265-21.ucode
iwlwifi-8265-22.ucode
iwlwifi-8265-27.ucode
iwlwifi-8265-31.ucode
```

The revision 27 seems to be loaded by 4.12. There aren't any updates for iwlwifi-8265 in neither linux-firmware nor iwlwifi's linux-firmware. :cry:

**Comment by bendlas** [comment]
I was thinking the same thing, but my hope was for the critical fix to not be in the ucode, but in the loading stub ...
I hope to be able to try this later in the evening

**Comment by NeQuissimus** [comment]
I am about to push the 4.13.4 kernel, it has a fix for something iwlwifi related, maybe that will fix this?!

https://cdn.kernel.org/pub/linux/kernel/v4.x/ChangeLog-4.13.4

**Comment by NeQuissimus** [comment]
Try with ab3ff4af9fb3a32888730496a1561290a2621433

**Comment by bendlas** [comment]
I finally got my wifi card working again, after the iwl release yesterday. Strange thing is: I had to enable `hardware.enableRedistributableFirmware`, for the updated `firmwareLinuxNonfree` package to even be built. I definitely haven't had to do this before. And I'm still not sure, if I even need the updated release. Did the default for this change, somehow?

What's the best course forward? Should we enable redistributable firmware by default?
